[Train][Templates] Add LoRA support to Llama-2 finetuning example (#37794)
ray-project/ray: doc/source/templates/04_finetuning_llms_with_deepspeed/README.md

### Launching fine-tuning

> The script is written using Ray Train + Deepspeed integration via accelerate API. The script is general enough that it can be used to fine-tune all released sizes of LLaMA 2 models. 

Reviewer: ```suggestion
The script is written using Ray Train + Deepspeed integration via accelerate API. The script is general enough that it can be used to fine-tune all released sizes of Llama-2 models. 
```

Reviewer: nit: Let's keep it consistent through out the readme?

| 70B        | `meta-llama/Llama-2-70b-hf`  | 8                     | 32x A10G (24G) | ~190 min.             |

To launch a successful finetuning you can use the following command:

```
./run_llama_ft.sh --size=7b --num-epochs=10
```

Note that the learning rate schedule scales with the number of epochs.

> Hence, training with ```./run_llama_7b.sh --num-epochs=2``` yields different results after training for 1 epoch then ```./run_llama_7b.sh --num-epochs=1```.

Reviewer: wait the script does not allow you to specify `--num-epochs`.Let's keep it simple and hardcode to 1 epoch total.

Reviewer: The point of this shell script is to set a hello-world entry point.

Reviewer: No need to highligh the complexities of learning rate.

### Launching LoRA fine-tuning

> You can utilize [LoRA](https://arxiv.org/abs/2106.09685) to achieve similar (or better) finetuning results than vanilla finetuning while using less compute during training and enabling quick swapping of finetuned parameters.

Reviewer: ```suggestion
You can utilize [LoRA](https://arxiv.org/abs/2106.09685) to achieve more resource efficient fine-tuning results than full-parameter fine-tuning.
```

The command for testing a lora fine-tuning is:

```

> python finetune_hf_llm.py --as-test --lora 

Reviewer: ```suggestion
./run_llama_ft.sh --size=7b --lora 
```

```

When adapting `target_modules` in the LoRA config (at lora_configs/lora.json), be aware that different models have different names for the
sub-models that can be tuned with LoRA. For example, LLaMa 2 requires {"target_modules": ["q_proj", "v_proj"]} and Falcon requires {"target_modules": ["query_key_value"]}. Running with `--as-test` will tell you what the submodules the model has.

> To launch a succesful LoRA finetuning, you can use the following command:
> 
> ```
> ./run_llama_ft.sh --size=7b --num-epochs=10 --lora
> ```

Reviewer: ```suggestion
```

Reviewer: we said it too many times what the command is :)

| 70B        | `meta-llama/Llama-2-70b-hf`  | 8                     | 32x A10G (24G) | ~190 min.             |

To launch a full fine-tuning you can use the following command:

```
./run_llama_ft.sh --size=7b
```

### Launching LoRA fine-tuning

You can utilize [LoRA](https://arxiv.org/abs/2106.09685) to achieve more resource efficient fine-tuning results than full-parameter fine-tuning, but unlocking smaller instance types and more effecient model serving.
To launch a LoRA fine-tuning, you can use the following command or similar commands for other model sizes:

```
./run_llama_ft.sh --size=7b --lora
```

Fine-tuning a model with LoRA results in a checkpoint containing only the fine-tuned weights.
As an example, the default Llama 2 LoRA configuration should yield a 42/64/202MB checkpoint for 7B/13B/70B models.
If we want to evaluate the model after training, we can merge the model weights with the original (non-fine-tuned) model.
We provide a script to merge the fine-tuned weights with the original weights to produce a full-parameter checkpoint.

> The script has high CPU memory requirements because it requires us to load all parameters into memory at the same time, 152GB for 7B/13B/70B models. Downloading and loading the original weights should take ~1min/~2min/~10min each on a p4de.24xlarge instance for the 7B/13B/70B models. You can run the script as follows:

Reviewer: 152GB for 7B/13B/70B ???

Author: Corrected, thanks!

```

> python merge_model_weights.py --model-name=7b --checkpoint=<path to your checkpoint> --output-path=<desired output path>

Reviewer: merge_lora_weights.py vs. merge_model_weights.py
